tree/20_cfe_basics/time_lib.cf

```diff
@@ -0,0 +1,203 @@
+      "splay_null" expression => strcmp("${splay}", "0");
+
+      # if splay >= run_periodicity -> error
+      # this must be checked in parent too if you have multiple periodicity
+      "job_${job_id}_error_splay_too_big" not => islessthan("${splay}", "${job_periodicity}"),
+                                          scope => "namespace",
+                                          ifvarclass => "!null_job_periodicity";
+
+      # if run_shift != 0 -> unpredictable
+      "unpredictable" not => strcmp("${run_shift}", "0");
+
+      # global error
+      "job_${job_id}_error" expression => "job_${job_id}_error_null_agent_periodicity|job_${job_id}_error_splay_too_small|job_${job_id}_error_splay_too_big",
+                            scope => "namespace";
+
+    any::
+      "noerror" expression => "!job_${job_id}_error";
+
+  reports:
+    unpredictable::
+      "Warning, there is a shift of ${run_shift} minutes between agent run period and ${job_id} run period";
+    splay_too_small.!splay_null::
+      "Warning, splay for ${job_id} is less than your agent run period but not null, it will be ignored";
+}
+
+
+# Simple Periodic Scheduler, it raises a class on a regular perioridicity, based on a user friendly definition.
+#
+# job_id: a string to identify this job
+# agent_periodicity: how often you run the agent in minutes
+# splay_minutes, splay_hours: on how much time you want to spread the job
+# start_on_minutes, start_on_hours, start_on_day_of_week: when should be the first run
+# periodicity_minutes, periodicity_hours, periodicity_days: how often should the job run
+# mode: "nodups": avoid duplicate runs in the same period
+#       "catchup": avoid duplicates and one or more run have been missed, run once before next period
+#       "stateless": no check is done on past runs
+bundle agent schedule_simple_generic(job_id, agent_periodicity,
+                                     splay_minutes, splay_hours,
+                                     start_on_minutes, start_on_hours, start_on_day_of_week,
+                                     periodicity_minutes, periodicity_hours, periodicity_days, mode)
+{
+  vars:
+    any::
+      # splay in minutes
+      "splay_real" string => eval("${splay_minutes} + 60 * ${splay_hours}", "math", "infix");
+      "splay" int => "${splay_real}";
+
+      # periodicity in minutes
+      "job_periodicity_real" string => eval("${periodicity_minutes} + 60 * ${periodicity_hours} + 60 * 24 * ${periodicity_days}", "math", "infix");
+      "job_periodicity" int => "${job_periodicity_real}";
+
+      # compute time between start_* and now
+      # (min_of_day - start_minutes) + 60 * (hour of day - start hours) + 60 * 24 * (day of week - start day)%7 
+      "now_real" string => eval(strftime("localtime", "%M - ${start_on_minutes} + 60 * (%H - ${start_on_hours}) + 60 * 24 * ((%u - ${start_on_day_of_week} + 7)%%7)", "${date.timestamp}"), "math", "infix");
+      "now" int => "${now_real}";
```

Comment:
now is a really poor name for that, it leads to believe that it is now, while is a difference between the start of job and now.